swift/Todo/Controller/TodoController.swift

```diff
@@ -0,0 +1,182 @@
+        do {
+            guard let taskDoc = try DB.shared.getTaskByID(id: taskID)
+            else {
+                fatalError("Couldn't fetch task with ID: \(taskID)")
+            }
+            return taskDoc
+        } catch {
+            fatalError("Error fetching task: \(error.localizedDescription)")
+        }
+    }
+    
+// - MARK: Common Task functions
+    
+    public static func generateTasks(taskListID: String, withPhoto: Bool, delegate: TaskControllerDelegate) {
+        do {
+            let taskListDoc = getTaskListDoc(fromID: taskListID)
+            try DB.shared.generateTasks(taskList: taskListDoc, numbers: 50, includesPhoto: withPhoto)
+        } catch {
+            delegate.presentError(error, message: "Error generating tasks, with photo: \(withPhoto)")
+        }
+    }
+    
+    public static func createTask(withName name: String, inTaskList taskListID: String, delegate: TaskControllerDelegate) {
+        do {
+            let taskListDoc = getTaskListDoc(fromID: taskListID)
+            try DB.shared.createTask(taskList: taskListDoc, task: name)
+        } catch {
+            delegate.presentError(error, message: "Couldn't create task")
+        }
+    }
+    
+    public static func deleteTask(taskID: String, delegate: TaskControllerDelegate) {
+        do {
+            try DB.shared.deleteTask(taskID: taskID)
+        } catch {
+            delegate.presentError(error, message: "Couldn't delete task")
+        }
+    }
+    
+    public static func updateTask(taskID: String, name: String, delegate: TaskControllerDelegate) {
+        do {
+            try DB.shared.updateTask(taskID: taskID, task: name)
+        } catch {
+            delegate.presentError(error, message: "Couldn't update task name")
+        }
+    }
+    
+    public static func updateTask(taskID: String, complete: Bool, delegate: TaskControllerDelegate) {
+        do {
+            try DB.shared.updateTask(taskID: taskID, complete: complete)
+        } catch {
+            delegate.presentError(error, message: "Couldn't update task completion")
+        }
+    }
+    
+    public static func updateTask(taskID: String, image: UIImage, delegate: TaskControllerDelegate) {
+        do {
+            try DB.shared.updateTask(taskID: taskID, image: image)
+        } catch {
+            delegate.presentError(error, message: "Couldn't update task image")
+        }
+    }
+    
+    public static func toggleTaskComplete(taskID: String, delegate: TaskControllerDelegate) {
+        let taskDoc = getTaskDoc(fromID: taskID)
+        let complete = taskDoc.boolean(forKey: "complete")
+        TodoController.updateTask(taskID: taskID, complete: !complete, delegate: delegate)
+    }
+    
+    public static func deleteImage(taskID: String) {
+        do {
+            try DB.shared.updateTask(taskID: taskID, image: nil)
+        } catch {
+            AppController.logger.log("Error deleting image: \(error.localizedDescription)")
```

Comment:
Is it possible to present the error as well?

Reply:
Good point. I did have the call to `deleteImage` asynchronous, but now I've taken that away in order to be able to present the error. The asynchronous probably didn't matter much anyway.